File: directions.py
class East:

  def __str__(self):
    return 'E'

  def left(self):
    return North()

  def right(self):
    return South()


class North:

  def __str__(self):
    return 'N'

  def left(self):
    return West()

  def right(self):
    return East()


class South:

  def __str__(self):
    return 'S'

  def left(self):
    return East()

  def right(self):
    return West()


class West:

  def __str__(self):
    return 'W'

  def left(self):
    return South()

  def right(self):
    return North()

File: directions.ts
export interface Direction {
  left(): Direction;
  right(): Direction;
  toString(): string;
}

export class East implements Direction {
  left(): Direction {
    return new North();
  }

  right(): Direction {
    return new South();
  }

  toString(): string {
    return 'E';
  }
}

export class North implements Direction {
  left(): Direction {
    return new West();
  }

  right(): Direction {
    return new East();
  }

  toString(): string {
    return 'N';
  }
}

export class South implements Direction {
  left(): Direction {
    return new East();
  }

  right(): Direction {
    return new West();
  }

  toString(): string {
    return 'S';
  }
}

export class West implements Direction {
  left(): Direction {
    return new South();
  }

  right(): Direction {
    return new North();
  }

  toString(): string {
    return 'W';
  }
}
